File: scripts/update_roster.py
import json,subprocess,sys
from datetime import datetime,timezone
from pathlib import Path
import requests

TEAM_ID=143
SEASON=datetime.now().year
OUT=Path("data/check-mlb-roster.json")
BASE=f"https://statsapi.mlb.com/api/v1/teams/{TEAM_ID}/roster"

def roster(t):
    r=requests.get(BASE,params={"rosterType":t,"season":SEASON},timeout=30)
    r.raise_for_status()
    return r.json().get("roster",[])

def val(row,*keys):
    for k in keys:
        if k in row and row[k] is not None:
            try:
                if row[k]!=row[k]: continue
            except: pass
            try:return row[k].item()
            except:return row[k]
    return None

try:
    from pybaseball import batting_stats,pitching_stats,statcast_batter_percentile_ranks,statcast_pitcher_percentile_ranks
except ImportError:
    subprocess.check_call([sys.executable,"-m","pip","install","pybaseball"])
    from pybaseball import batting_stats,pitching_stats,statcast_batter_percentile_ranks,statcast_pitcher_percentile_ranks

r40=roster("40Man")
active={x["person"]["id"] for x in roster("active")}
players=[]

for x in r40:
    p=x["person"]; pid=p["id"]; il=False
    try:
        q=requests.get(f"https://statsapi.mlb.com/api/v1/people/{pid}",params={"hydrate":"rosterEntries"},timeout=30).json()["people"][0]
        il=any(
            str(e.get("status",{}).get("code","")).upper() in {"IL","10DIL","15DIL","60DIL"}
            or "injured list" in str(e.get("status",{}).get("description","")).lower()
            for e in q.get("rosterEntries",[])
        )
    except: pass

    info={}
    try:
        q=requests.get(f"https://statsapi.mlb.com/api/v1/people/{pid}",timeout=30).json()["people"][0]
        b=q.get("birthDate"); age=None
        if b:
            d=datetime.strptime(b,"%Y-%m-%d").date(); t=datetime.now().date()
            age=t.year-d.year-((t.month,t.day)<(d.month,d.day))
        pos=q.get("primaryPosition",{}).get("abbreviation")
        if pos not in {"P","C","1B","2B","3B","SS","LF","CF","RF"}: pos=x.get("position",{}).get("abbreviation")
        bat=q.get("batSide",{}).get("code"); throw=q.get("pitchHand",{}).get("code")
        info={"name":q.get("fullName"),"number":q.get("primaryNumber") or x.get("jerseyNumber"),"position":pos,"B/T":f"{bat}/{throw}" if bat and throw else None,"height":q.get("height"),"weight":q.get("weight"),"birthDate":b,"age":age,"birthCountry":q.get("birthCountry")}
    except:
        info={"name":p.get("fullName"),"number":x.get("jerseyNumber"),"position":x.get("position",{}).get("abbreviation"),"B/T":None,"height":None,"weight":None,"birthDate":None,"age":None,"birthCountry":None}

    players.append({"id":pid,"rosterClass":"IL" if il else "ACTIVE" if pid in active else "40-MAN","is40Man":True,"isActive":pid in active,"isIL":il,"bio":info})

try:b=batting_stats(SEASON,SEASON,qual=0)
except:b=None
try:p=pitching_stats(SEASON,SEASON,qual=0)
except:p=None
try:bp=statcast_batter_percentile_ranks(SEASON)
except:bp=None
try:pp=statcast_pitcher_percentile_ranks(SEASON)
except:pp=None

def row(df,pid):
    if df is None:return None
    for c in ("MLBAMID","player_id","mlb_id"):
        if c in df.columns:
            z=df[df[c]==pid]
            if not z.empty:return z.iloc[0]

for x in players:
    pid=x["id"]; br=row(b,pid); pr=row(p,pid); bs=row(bp,pid); ps=row(pp,pid)
    x["batting"]={k:val(br,*v) if br is not None else None for k,v in {
        "G":("G",),"PA":("PA",),"H":("H",),"HR":("HR",),"RBI":("RBI",),
        "AVG":("AVG",),"OPS":("OPS",),"BABIP":("BABIP",),"fWAR":("WAR",),
        "wRC+":("wRC+",),"wOBA":("wOBA",),"DRS":("DRS",),
        "OAA":("OAA","oaa")}.items()}
    x["pitching"]={k:val(pr,*v) if pr is not None else None for k,v in {
        "G":("G",),"IP":("IP",),"HR":("HR",),"BB":("BB",),"K":("SO","K"),
        "fWAR":("WAR",),"FIP":("FIP",),"xFIP":("xFIP",),
        "K-BB%":("K-BB%",),"ERA":("ERA",)}.items()}
    w=val(pr,"W") if pr is not None else None;l=val(pr,"L") if pr is not None else None
    x["pitching"]["W-L"]=f"{w}-{l}" if w is not None and l is not None else None
    x["pitching"]["Whiff%"]=val(ps,"whiff_percent","Whiff%") if ps is not None else None

OUT.parent.mkdir(exist_ok=True)
OUT.write_text(json.dumps({"team":"Philadelphia Phillies","teamId":TEAM_ID,"season":SEASON,"updatedAt":datetime.now(timezone.utc).isoformat(),"count":len(players),"players":players},ensure_ascii=False,indent=2),encoding="utf-8")

File: scripts/update_roster.mjs
const TEAM_ID = 143;
const SEASON = new Date().getFullYear();
const API = "https://statsapi.mlb.com/api/v1";

const fs = await import("node:fs/promises");
const path = await import("node:path");

const DATA_DIR = path.join(process.cwd(), "data");
const OUTPUT = path.join(DATA_DIR, "players.json");

async function getJSON(url) {
  const res = await fetch(url, {
    headers: {
      "User-Agent": "Phillies-App/1.0"
    }
  });

  if (!res.ok) {
    throw new Error(`${res.status} ${res.statusText}: ${url}`);
  }

  return res.json();
}

function value(obj, key) {
  return obj && obj[key] !== undefined ? obj[key] : null;
}

function classifyStatus(entry) {
  const code =
    entry?.status?.code ||
    entry?.status?.type ||
    entry?.status?.description ||
    "";

  const text = String(code).toLowerCase();

  if (
    text.includes("il") ||
    text.includes("injured") ||
    text.includes("10-day") ||
    text.includes("15-day") ||
    text.includes("60-day") ||
    text.includes("7-day")
  ) {
    return "IL";
  }

  if (
    text === "a" ||
    text.includes("active") ||
    text.includes("active roster")
  ) {
    return "ACTIVE";
  }

  return "40-MAN";
}

function isPitcher(player) {
  return player?.primaryPosition?.abbreviation === "P";
}

function isHitter(player) {
  return !isPitcher(player);
}

function findStat(people, type) {
  const stats = people?.stats || [];

  for (const group of stats) {
    if (
      group?.type?.displayName?.toLowerCase() ===
      type.toLowerCase()
    ) {
      const split = group?.splits?.[0];
      if (split?.stat) return split.stat;
    }

    if (
      group?.type?.name?.toLowerCase() ===
      type.toLowerCase()
    ) {
      const split = group?.splits?.[0];
      if (split?.stat) return split.stat;
    }
  }

  return null;
}

function pick(source, fields) {
  if (!source) {
    return Object.fromEntries(
      fields.map(key => [key, null])
    );
  }

  return Object.fromEntries(
    fields.map(key => [key, value(source, key)])
  );
}

function getValue(source, keys) {
  if (!source) return null;

  for (const key of keys) {
    if (
      source[key] !== undefined &&
      source[key] !== null
    ) {
      return source[key];
    }
  }

  return null;
}

function buildHitting(
  season,
  advanced,
  sabermetrics
) {
  return {
    season: pick(season, [
      "gamesPlayed",
      "plateAppearances",
      "atBats",
      "runs",
      "hits",
      "doubles",
      "triples",
      "homeRuns",
      "rbi",
      "baseOnBalls",
      "strikeOuts",
      "stolenBases",
      "caughtStealing",
      "avg",
      "obp",
      "slg",
      "ops",
      "babip",
      "totalBases"
    ]),

    advanced: pick(advanced, [
      "iso",
      "pitchesPerPlateAppearance",
      "walksPerPlateAppearance",
      "strikeoutsPerPlateAppearance",
      "homeRunsPerPlateAppearance",
      "walksPerStrikeout",
      "extraBaseHits"
    ]),

    sabermetrics: pick(sabermetrics, [
      "war",
      "wpa",
      "wpaMinus",
      "wpaPlus",
      "re24",
      "wrcPlus",
      "wraa",
      "woba"
    ])
  };
}

function buildPitching(
  season,
  advanced,
  sabermetrics
) {
  return {
    season: pick(season, [
      "gamesPlayed",
      "gamesStarted",
      "wins",
      "losses",
      "saves",
      "holds",
      "inningsPitched",
      "hits",
      "runs",
      "earnedRuns",
      "homeRuns",
      "baseOnBalls",
      "strikeOuts",
      "era",
      "whip"
    ]),

    advanced: pick(advanced, [
      "strikeoutsPer9Inn",
      "walksPer9Inn",
      "hitsPer9Inn",
      "homeRunsPer9",
      "strikeoutWalkRatio",
      "strikePercentage"
    ]),

    sabermetrics: {
      war: getValue(sabermetrics, [
        "war"
      ]),

      fip: getValue(sabermetrics, [
        "fip"
      ]),

      xfip: getValue(sabermetrics, [
        "xfip"
      ]),

      "k-bb%": getValue(sabermetrics, [
        "k-bb%",
        "kBB%",
        "kbb%",
        "kBBPct",
        "kBbPct",
        "kBBPercent",
        "kbbPercent"
      ]),

      wpa: getValue(sabermetrics, [
        "wpa"
      ]),

      re24: getValue(sabermetrics, [
        "re24"
      ])
    }
  };
}

async function main() {
  console.log(
    `Updating Phillies players: ${SEASON}`
  );

  const rosterURL =
    `${API}/teams/${TEAM_ID}/roster` +
    `?rosterType=40Man&season=${SEASON}`;

  const rosterData = await getJSON(rosterURL);
  const roster = rosterData.roster || [];

  if (!roster.length) {
    throw new Error(
      "40-man roster is empty. Existing data will not be overwritten."
    );
  }

  const players = [];

  for (const entry of roster) {
    const personId = entry?.person?.id;

    if (!personId) continue;

    try {
      const personURL =
        `${API}/people/${personId}` +
        `?season=${SEASON}` +
        `&hydrate=` +
        `stats(` +
        `group=[hitting,pitching],` +
        `type=[season,seasonAdvanced,sabermetrics],` +
        `season=${SEASON}` +
        `)`;

      const personData = await getJSON(personURL);
      const person = personData?.people?.[0];

      if (!person) continue;

      const seasonHitting =
        findStat(person, "season");

      const advancedHitting =
        findStat(person, "seasonAdvanced");

      const sabermetricsHitting =
        findStat(person, "sabermetrics");

      const groups = person?.stats || [];

      function getGroupStat(
        groupName,
        typeName
      ) {
        const group = groups.find(
          g =>
            String(
              g?.group?.displayName ||
              g?.group?.name ||
              ""
            ).toLowerCase() ===
              groupName.toLowerCase() &&
            String(
              g?.type?.displayName ||
              g?.type?.name ||
              ""
            ).toLowerCase() ===
              typeName.toLowerCase()
        );

        return group?.splits?.[0]?.stat || null;
      }

      const hittingSeason =
        getGroupStat(
          "hitting",
          "season"
        ) || seasonHitting;

      const hittingAdvanced =
        getGroupStat(
          "hitting",
          "seasonAdvanced"
        ) || advancedHitting;

      const hittingSabermetrics =
        getGroupStat(
          "hitting",
          "sabermetrics"
        ) || sabermetricsHitting;

      const pitchingSeason =
        getGroupStat(
          "pitching",
          "season"
        );

      const pitchingAdvanced =
        getGroupStat(
          "pitching",
          "seasonAdvanced"
        );

      const pitchingSabermetrics =
        getGroupStat(
          "pitching",
          "sabermetrics"
        );

      const pitcher = isPitcher(person);

      const player = {
        id: person.id,

        name:
          person.fullName ?? null,

        firstName:
          person.firstName ?? null,

        lastName:
          person.lastName ?? null,

        number:
          person.primaryNumber ??
          entry.jerseyNumber ??
          null,

        position: {
          code:
            person.primaryPosition?.code ??
            null,

          name:
            person.primaryPosition?.name ??
            null,

          abbreviation:
            person.primaryPosition
              ?.abbreviation ?? null
        },

        bt: {
          bats:
            person.batSide?.code ??
            null,

          throws:
            person.pitchHand?.code ??
            null
        },

        physical: {
          height:
            person.height ?? null,

          weight:
            person.weight ?? null,

          birthDate:
            person.birthDate ?? null,

          birthCity:
            person.birthCity ?? null,

          birthCountry:
            person.birthCountry ?? null
        },

        roster: {
          status:
            classifyStatus(entry),

          rosterType:
            "40-MAN",

          active:
            classifyStatus(entry) ===
            "ACTIVE",

          injured:
            classifyStatus(entry) ===
            "IL",

          fortyMan: true
        },

        stats: pitcher
          ? {
              type: "pitcher",

              pitching:
                buildPitching(
                  pitchingSeason,
                  pitchingAdvanced,
                  pitchingSabermetrics
                ),

              hitting: null
            }
          : {
              type: "hitter",

              hitting:
                buildHitting(
                  hittingSeason,
                  hittingAdvanced,
                  hittingSabermetrics
                ),

              pitching: null
            },

        updatedAt:
          new Date().toISOString()
      };

      players.push(player);

      console.log(
        `${player.name} -> ` +
        `${player.roster.status} -> ` +
        `${player.stats.type}`
      );

    } catch (error) {
      console.error(
        `Failed player ${personId}:`,
        error.message
      );
    }
  }

  if (!players.length) {
    throw new Error(
      "No players collected. Existing data will not be overwritten."
    );
  }

  const output = {
    team: {
      id: TEAM_ID,
      name: "Philadelphia Phillies"
    },

    season: SEASON,

    updatedAt:
      new Date().toISOString(),

    source:
      "MLB Stats API",

    players
  };

  await fs.mkdir(
    DATA_DIR,
    { recursive: true }
  );

  await fs.writeFile(
    OUTPUT,
    JSON.stringify(
      output,
      null,
      2
    ),
    "utf8"
  );

  console.log(
    `Saved ${players.length} players.`
  );

  console.log(OUTPUT);
}

main().catch(error => {
  console.error(error);
  process.exit(1);
});
